src/scales/scale.time.js

```diff
@@ -183,85 +169,159 @@ function parse(scale, input) {
+function updateMajors(scale, majors, unit) {
+	const units = scale._units;
+	const time = majors[unit] + 1000; // make sure we get the next major, even if matching
+	for (let i = units.indexOf(unit); i > 0; --i) {
+		const majorUnit = units[i];
+		majors[majorUnit] = roundUp(scale, time, majorUnit, true);
+	}
 }
 
-function determineMajorUnit(unit) {
-	for (var i = UNITS.indexOf(unit) + 1, ilen = UNITS.length; i < ilen; ++i) {
-		if (INTERVALS[UNITS[i]].common) {
-			return UNITS[i];
+function getUnit(time, units, majors) {
+	for (let i = units.length - 1; i > 0; --i) {
+		const unit = units[i];
+		if (time >= majors[unit]) {
+			return unit;
 		}
 	}
+	return units[0];
 }
 
-/**
- * Generates a maximum of `capacity` timestamps between min and max, rounded to the
- * `minor` unit using the given scale time `options`.
- * Important: this method can return ticks outside the min and max range, it's the
- * responsibility of the calling code to clamp values if needed.
- */
-function generate(scale, min, max, capacity) {
-	const adapter = scale._adapter;
-	const options = scale.options;
-	const timeOpts = options.time;
-	const minor = timeOpts.unit || determineUnitForAutoTicks(timeOpts.minUnit, min, max, capacity);
-	const stepSize = resolve([timeOpts.stepSize, timeOpts.unitStepSize, 1]);
-	const weekday = minor === 'week' ? timeOpts.isoWeekday : false;
-	const ticks = [];
-	let first = min;
-	let time;
+function getFirstAndLast(scale, extendBounds) {
+	const {_adapter, _stepSize, _units, min, max} = scale;
+	const minor = _units[0];
+	const firstMethod = extendBounds ? roundDown : roundUp;
+	const lastMethod = extendBounds ? roundUp : roundDown;
+	let first = scale._userMin || firstMethod(scale, min, minor);
+	let last = scale._userMax || lastMethod(scale, max, minor);
 
-	// For 'week' unit, handle the first day of week option
-	if (weekday) {
-		first = +adapter.startOf(first, 'isoWeek', weekday);
+	if (_stepSize > 1) {
+		last = _adapter.add(first, _stepSize * Math.round(_adapter.diff(last, first, minor) / _stepSize), minor);
 	}
 
-	// Align first ticks on unit
-	first = +adapter.startOf(first, weekday ? 'day' : minor);
-
-	// Prevent browser from freezing in case user options request millions of milliseconds
-	if (adapter.diff(max, min, minor) > 100000 * stepSize) {
-		throw min + ' and ' + max + ' are too far apart with stepSize of ' + stepSize + ' ' + minor;
+	if (_units.length > 1 && _stepSize > 1) {
+		let major = firstMethod(scale, min, _units[1]);
+		let diff = _adapter.diff(Math.max(first, major), Math.min(first, major), minor);
+		if (diff <= _stepSize) {
+			first = major;
+		}
+		major = lastMethod(scale, last, _units[1]);
+		diff = _adapter.diff(Math.max(last, major), Math.min(last, major), minor);
+		if (diff <= _stepSize) {
+			last = major;
+		}
 	}
 
-	for (time = first; time < max; time = +adapter.add(time, stepSize, minor)) {
-		ticks.push(time);
+	return {first, last};
+}
+
+function addTick(scale, tick, ticks, positions) {
```

Comment:
I don't quite understand what this method's doing. There's lots of hardcoded values in it like `-0.1`, `1.1`, `0.8`, etc. that don't have any meaning to me. Is this doing something fundamentally different than the old `_getLabelCapacity`?

Reply:
Yes, this is a bit different. Using the lookup table's range [0..1] to try and generate evenly spaced ticks. So the (abstract length) is divided in slots of how many ticks we are trying to fit, and within those slots we are favoring the major units. The `-0.1` and `1.1` are to allow range extension by one tick. Just put something there to get it working, that should be `1/ maxNumOfTicks` below `0` and above `1` to work correctly with large number of ticks too. 

The 0.8 is a experimental value determining how close to the slot center a tick must be, for it to be approved.

Its still a draft, and I'm not quite happy with it in all situations (at least not yet).